dumper/app/src/test/java/com/google/edwmigration/dumper/application/dumper/connector/oracle/AbstractOracleConnectorTest.java

```diff
@@ -44,4 +50,52 @@ public void buildProperties_success() {
     assertEquals(EXAMPLE_PASSWORD, actual.getProperty("password"));
     assertEquals(EXAMPLE_USER, actual.getProperty("user"));
   }
+
+  @Test
+  public void buildUrl_providedBoth_throwsException() {
+    when(arguments.getOracleServicename()).thenReturn("ORCLPDB");
+    when(arguments.getOracleSID()).thenReturn("ORCLPDB1");
+    // Act
+    ThrowingRunnable runnable = () -> AbstractOracleConnector.buildUrl(arguments);
+    // Assert
+    assertThrows(MetadataDumperUsageException.class, runnable);
+  }
+
+  @Test
+  public void buildUrl_providedNone_throwsException() {
+    ThrowingRunnable runnable = () -> AbstractOracleConnector.buildUrl(arguments);
+    assertThrows(MetadataDumperUsageException.class, runnable);
+  }
+
+  @Test
+  public void buildUrl_providedServiceName_success() {
+    when(arguments.getOracleServicename()).thenReturn("ORCLPDB");
+    when(arguments.getHost()).thenReturn("localhost");
+    when(arguments.getPort(anyInt())).thenReturn(1521);
+    // Act
+    String url = AbstractOracleConnector.buildUrl(arguments);
+    // Assert
+    assertEquals("jdbc:oracle:thin:@//localhost:1521/ORCLPDB", url);
+  }
+
+  @Test
+  public void buildUrl_providedSid_success() {
+    when(arguments.getOracleSID()).thenReturn("ORCLPDB1");
+    when(arguments.getHost()).thenReturn("localhost");
+    when(arguments.getPort(anyInt())).thenReturn(1521);
+    // Act
+    String url = AbstractOracleConnector.buildUrl(arguments);
+    // Assert
+    assertEquals("jdbc:oracle:thin:@localhost:1521:ORCLPDB1", url);
+  }
+
+  @Test
+  public void buildUrl_providedUrl_success() {
```

Comment:
I would suggest adding also a test case, when both URL and host command-line flags are provided.

Reply:
Thanks for bringing this up, it seems like an important corner case. I'm not sure what's the correct behavior here.

The current approach is to silently ignore the host, which I don't like, because a) it's bad to silently ignore a flag, b) the reason why we choose the url and not the host is unclear to the user and c) we should try to limit the number of possible flag combinations as much as possible to reduce complexity.

Ideally, I'd like to throw an usage exception, but in any case I'd discuss this further before including it in tests.